Hymns/Home/HomeView.swift

```diff
@@ -2,29 +2,40 @@ import Resolver
 import SwiftUI
 
 struct HomeView: View {
+
     @State var searchText: String = ""
+    @ObservedObject private var viewModel: HomeViewModel
+
     var allHymns: [DummyHymnView] = testData
 
+    init(viewModel: HomeViewModel) {
+        self.viewModel = viewModel
+    }
+
     var body: some View {
         NavigationView {
             VStack {
                 NavigateToSearch(searchText: searchText) //passes necessary param to extracted subview
 
-                //This list does two things. It filters all the Hymns based on search function and it also provides the linking to their respective detailed views.
                 List {
-                    ForEach(self.allHymns.filter { self.searchText.isEmpty ? true : $0.songTitle.contains(self.searchText)}) { hymn in
-                        NavigationLink(destination: HymnLyricsView(viewModel: Resolver.resolve())) {
-                            Text(hymn.songTitle)}
+                    ForEach(self.viewModel.recentSongs) { recentSong in
+                        NavigationLink(destination: recentSong.destinationView) {
+                            SongResultView(viewModel: recentSong)
+                        }
```

Comment:
So let me see if I understand this, by calling recentSong.destinationView we are going through our HomeViewModel published variable recentSong to then use the SongResultViewModel which then by virtue of being of a generic type that conforms to a view. It is able to be set as a destination and brings us to the HymnLyricsView for that given hymn.... My head is spinning trying to understand this thing. Where does the SongResultView come into play?

Reply:
Yeah, you're right.

So here's HomeViewModel:
`class HomeViewModel: ObservableObject {

    @Published var recentSongs: [SongResultViewModel<HymnLyricsView>] = [SongResultViewModel<HymnLyricsView>]()

    init(recentSongs: [SongResultViewModel<HymnLyricsView>]) {
        // TODO fetch recent songs instead of passing it in
        self.recentSongs = recentSongs
    }
}`

Here, we're passing in the SongResultViewModel because we don't have any data source yet. But in the future, we'd be making a call to the database to get the song results, and we'd create the SongResultViewModel based on the song result model from the db.

An example would be:
`    init() {
        var dbModels: RecentSongDatabseModel = database.getRecentSongs()
        recentSongs = dbModels.map(SongResultViewModel)
    }`

So you get the db models, and create the data model here, which contains the navigation information

Reply:
Feel free to ping me on messenger if you want me to explain it more. Happy to hop on a call or explain over chat if that wasn't clear